includes/class-llms-prevent-concurrent-logins.php

```diff
@@ -0,0 +1,165 @@
+<?php
+/**
+ * LLMS_Prevent_Concurrent_Logins class file
+ *
+ * @package LifterLMS/Classes
+ *
+ * @since [version]
+ * @version [version]
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * LLMS_Prevent_Concurrent_Logins class.
+ *
+ * @since [version]
+ */
+class LLMS_Prevent_Concurrent_Logins {
+
+	use LLMS_Trait_Singleton;
+
+	/**
+	 * Array of sessions for the current user.
+	 *
+	 * @var array
+	 */
+	private $user_sessions;
+
+	/**
+	 * Current user ID.
+	 *
+	 * @var int
+	 */
+	private $user_id;
+
+	/**
+	 * Private Constructor.
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	private function __construct() {
+
+		if ( get_option( 'lifterlms_prevent_concurrent_logins', false ) &&
+				! empty( get_option( 'lifterlms_prevent_concurrent_logins_roles', array() ) ) ) {
+
+			add_action( 'init', array( $this, 'init' ) );
+			add_action( 'init', array( $this, 'maybe_prevent_concurrent_logins' ) );
+		}
+
+	}
+
+	/**
+	 * Initialize.
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	public function init() {
+
+		$this->user_id = get_current_user_id();
+
+		if ( empty( $this->user_id ) ) {
+			return;
+		}
+
+		$this->user_sessions = wp_get_all_sessions();
+
+	}
+
+	/**
+	 * Maybe prevent current logins.
+	 *
+	 * @since [version]
+	 *
+	 * @return bool `true` if concurrent login prevented, `false` otherwise.
+	 */
+	public function maybe_prevent_concurrent_logins() {
+
+		// No logged in user or current user has only one active session: nothing to do.
+		if ( empty( $this->user_sessions ) || count( $this->user_sessions ) < 2 ) {
+			return false;
+		}
+
+		// Current user doesn't have any restricted role: nothing to do.
+		if ( empty( array_intersect( get_userdata( $this->user_id )->roles, get_option( 'lifterlms_prevent_concurrent_logins_roles', array() ) ) ) ) {
+			return false;
+		}
+
+		/**
+		 * Filters whether or not allowing a specific user to have concurrent sessions.
+		 *
+		 * @since [version]
+		 *
+		 * @param bool $allow   Whether or not the user should be allowed to have concurrent sessions.
+		 * @param int  $user_id WP_User ID of the current use.
+		 */
+		if ( (bool) apply_filters( 'llms_allow_user_concurrent_logins', false, $this->user_id ) ) {
+			return false;
+		}
+
+		// Current user doesn't have any restricted role: nothing to do.
+		if ( empty( array_intersect( get_userdata( $this->user_id )->roles, get_option( 'lifterlms_prevent_concurrent_logins_roles', array() ) ) ) ) {
```

Comment:
Just in case... I've seen people empty options and create fatal errors with code like this (that we then have to deal with)... Something I remember with admin notifications a while back, for example...

```suggestion
		if ( empty( array_intersect( get_userdata( $this->user_id )->roles, (array) get_option( 'lifterlms_prevent_concurrent_logins_roles', array() ) ) ) ) {
```